pilot/pkg/serviceregistry/kube/controller/ambient/sidecar_interop.go

```diff
@@ -0,0 +1,173 @@
+	// When sending information to a workload, we need two bits of information:
+	// * Does it support tunnel? If so, we will need to use HBONE
+	// * Does it have a waypoint? if so, we will need to send to the waypoint
+	// Record both of these.
+	// Note: currently, EDS uses the waypoint VIP for workload waypoints. This is probably something that will change.
+	WorkloadEds := krt.NewCollection(
+		Workloads,
+		func(ctx krt.HandlerContext, wl model.WorkloadInfo) *workloadEDS {
+			if wl.Waypoint == nil && wl.TunnelProtocol == workloadapi.TunnelProtocol_NONE {
+				return nil
+			}
+			res := &workloadEDS{
+				WorkloadKey:      wl.ResourceName(),
+				TunnelProtocol:   wl.TunnelProtocol,
+				ServiceHostnames: slices.Sort(maps.Keys(wl.Services)),
+			}
+			if wl.Waypoint != nil {
+				if wl.Waypoint.GetHostname() == nil {
+					// TODO!!!!!
+					panic("todo")
+				}
+				hn := wl.Waypoint.GetHostname()
+				waypointServiceKey := hn.Namespace + "/" + hn.Hostname
+				workloads := krt.Fetch(ctx, Workloads, krt.FilterIndex(WorkloadsByServiceKey, waypointServiceKey))
+				res.WaypointInstance = slices.Map(workloads, func(e model.WorkloadInfo) *workloadapi.Workload {
+					return e.Workload
+				})
+			}
+			return res
+		},
+		krt.WithName("WorkloadEds"))
+	ServiceEds := krt.NewCollection(
+		Services,
+		func(ctx krt.HandlerContext, svc model.ServiceInfo) *serviceEDS {
+			if svc.Service.Waypoint == nil {
+				return nil
+			}
+			if svc.Service.Waypoint.GetHostname() == nil {
+				// TODO!!!!!
+				panic("todo")
+			}
+			hn := svc.Service.Waypoint.GetHostname()
+			waypointServiceKey := hn.Namespace + "/" + hn.Hostname
+			workloads := krt.Fetch(ctx, Workloads, krt.FilterIndex(WorkloadsByServiceKey, waypointServiceKey))
+			return &serviceEDS{
+				ServiceKey: svc.ResourceName(),
+				WaypointInstance: slices.Map(workloads, func(e model.WorkloadInfo) *workloadapi.Workload {
+					return e.Workload
+				}),
+			}
+		},
+		krt.WithName("ServiceEds"))
+	WorkloadEds.RegisterBatch(
+		PushAllXds(xdsUpdater, func(i workloadEDS, updates sets.Set[model.ConfigKey]) {
+			for _, svc := range i.ServiceHostnames {
+				ns, hostname, _ := strings.Cut(svc, "/")
+				updates.Insert(model.ConfigKey{Kind: kind.ServiceEntry, Name: hostname, Namespace: ns})
+			}
+		}), false)
+	ServiceEds.RegisterBatch(
+		PushAllXds(xdsUpdater, func(svc serviceEDS, updates sets.Set[model.ConfigKey]) {
+			ns, hostname, _ := strings.Cut(svc.ServiceKey, "/")
+			updates.Insert(model.ConfigKey{Kind: kind.ServiceEntry, Name: hostname, Namespace: ns})
+		}), false)
+}
+
+func PushAllXds[T any](xds model.XDSUpdater, f func(T, sets.Set[model.ConfigKey])) func(events []krt.Event[T], initialSync bool) {
+	return func(events []krt.Event[T], initialSync bool) {
+		cu := sets.New[model.ConfigKey]()
+		for _, e := range events {
+			for _, i := range e.Items() {
+				f(i, cu)
+			}
+		}
+		if len(cu) == 0 {
+			return
+		}
+		xds.ConfigUpdate(&model.PushRequest{
+			Full:           false,
```

Comment:
This is just reminding me of the other open issue about removing full pushes. If a function called "PushAllXds" doesn't result in a full push, I don't know what should